### ISSUE TITLE
Virtualize QueryList items

### ISSUE THREAD

**Issue by frabarz** [palantir/blueprint#2632]
#### Environment

- __Package version(s)__: core@1.35.7 / labs@0.14.4
- __Browser and OS versions__: Opera 53, Manjaro Linux XFCE

#### Feature request

It would be great to have the `Select` list elements, or the `QueryList` for that matter, be virtualized to let it handle very long lists.

#### Examples

I wanted to use `Select` to handle a list with ~1400 elements, with just a `span` inside, and as it renders all the items the list takes a good while to appear and filter. I ended up piecing something with `react-tiny-virtual-list` and `Popover2`, but without a bunch of the capabilities of `Select`.

**Comment by giladgray** [comment]
Oh also you'll need to be on 2.x or ideally 3.0 for this

**Comment by cmslewis** [comment]
@adidahiya Thoughts on re-opening this? Using a custom `itemListRenderer` is actually non-trivial or impossible with the new "Create item" option. We may just need to improve the support story there, but I also think it's reasonable for the Select components to handle rendering virtualization internally.

**Comment by adidahiya** [comment]
Might consider https://github.com/clauderic/react-tiny-virtual-list for this

**Comment by gianpaj** [comment]
I agree this not being trivial. I'm having issues using `react-tiny-virtual-list`. Could anybody provide an example or help me figure this out?:

https://codesandbox.io/s/p3j721n57m

<img width="589" alt="Screenshot 2019-04-28 at 23 09 10" src="https://user-images.githubusercontent.com/899175/56869546-b4613f00-6a0a-11e9-99a8-35a91dd5bc1b.png">

I believe it's something to do with the `style` parameter of `renderItem` from `VirtualList`.

```jsx
  renderSelectList = ({
    items,
    itemsParentRef, // do i need this??
    renderItem,
  }: {
    items: Films.IFilm[],
    itemsParentRef: (ref: HTMLElement | null) => void,
    renderItem: (item: any, index: number) => JSX.Element | null,
  }) => {
    return (
      <VirtualList
        width="100%"
        height={400}
        itemCount={items.length}
        itemSize={20} // Also supports variable heights (array or function getter)
        renderItem={({ index, style }) => renderItem(items[index], index)}
      />
    );
  };
```

**Comment by cmslewis** [comment]
@gianpaj Would a `list-style: none` CSS rule come to the rescue here? The menu items look okay but for the erroneous bullets. Having not used `react-tiny-virtual-list` myself before, I'm not sure which selector would need that rule, but hopefully some quick DOM spelunking can help identify the right element.

**Comment by gianpaj** [comment]
Hi @cmslewis , thanks for that but actually the CSS is not the only issue.
The biggest issue is that the rest of the items are not appearing on the screen as I believe their position is not being updated when you're scrolling.
i.e. after 79. A Clockwork Orange, more films should appear.

Check the example: https://codesandbox.io/s/p3j721n57m
and gif: 
![May-09-2019 23-06-12](https://user-images.githubusercontent.com/899175/57483258-180f2600-72af-11e9-94a3-b3e80c91d8bc.gif)

That's my guess anyway

**Comment by saruman1234** [comment]
> I believe it's something to do with the `style` parameter of `renderItem` from `VirtualList`.
> 
> ```js-jsx
>   renderSelectList = ({
>     items,
>     itemsParentRef, // do i need this??
>     renderItem,
>   }: {
>     items: Films.IFilm[],
>     itemsParentRef: (ref: HTMLElement | null) => void,
>     renderItem: (item: any, index: number) => JSX.Element | null,
>   }) => {
>     return (
>       <VirtualList
>         width="100%"
>         height={400}
>         itemCount={items.length}
>         itemSize={20} // Also supports variable heights (array or function getter)
>         renderItem={({ index, style }) => renderItem(items[index], index)}
>       />
>     );
>   };
> ```

I think you're right, as I'm having a similar problem with a different virtualized list library.

Many of the virtualized/window libraries work by setting absolute positions on the elements in the list as you scroll. [This blank list effect is a very common scenario](https://github.com/bvaughn/react-window#why-is-my-list-blank-when-i-scroll))

Because the `renderItem(..)` method does not accept them, we cannot pass the `style` props through to the `<MenuItem/>` component being rendered.

**Comment by Reggino** [comment]
@saruman1234 That helped a lot. Thanks for sharing.

**Comment by Lonli-Lokli** [comment]
It would be great to revisit support for it
